### PR TITLE
add promise.any

### LINKED ISSUE #33843: proposal-promise-any
## Search Terms
tc39

## Suggestion
https://tc39.es/proposal-promise-any/ is stage-3 now

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #33843 

BTW: https://github.com/microsoft/TypeScript/pull/31085

### CODE REVIEW

File: src/lib/esnext.promise.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+declare var AggregateError: AggregateErrorConstructor;
+
+/**
+ * Represents the completion of an asynchronous operation
+ */
+interface PromiseConstructor {
+    /**
+     * The any function returns a promise that is fulfilled by the first given promise to be fulfilled, or rejected with an array of rejection reasons if all of the given promises are rejected. It resolves all elements of the passed iterable to promises as it runs this algorithm.
```
Comment (chicoxyzzy) [inline src/lib/esnext.promise.d.ts]: when it's rejecting, it's reject with `AggregateError`, not an array of rejection reasons

Comment (sandersn) [reply src/lib/esnext.promise.d.ts]: :+1:

```suggestion
     * The any function returns a promise that is fulfilled by the first given promise to be fulfilled, or rejected with an AggregateError containing an array of rejection reasons if all of the given promises are rejected. It resolves all elements of the passed iterable to promises as it runs this algorithm.
```

**Comment by sandersn** [comment]
A baseline needed to be updated after the merge from master. I'll merge this after CI finishes running.

**Comment by mt-yannick** [comment]
Hi, I am able to compile successfully but at runtime, I receive a `Promise.any` is undefined error. I am using typescript `3.9.5` in my project. This is my tsconfig:

```
{
  "compilerOptions": {
    "target": "ES6",
    "module": "commonjs",
    "strict": true,
    "esModuleInterop": true,
    "outDir": "dist",
    "types": ["puppeteer"],
    "lib": [
      "dom",
      "esnext",
      "esnext.promise"
    ],
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "forceConsistentCasingInFileNames": true,
    "inlineSourceMap": true,
  }
}
```

Any feedback would be much appreciated.

**Comment by alongite** [comment]
> Hi, I am able to compile successfully but at runtime, I receive a `Promise.any` is undefined error. I am using typescript `3.9.5` in my project. This is my tsconfig:
> 
> ```
> {
>   "compilerOptions": {
>     "target": "ES6",
>     "module": "commonjs",
>     "strict": true,
>     "esModuleInterop": true,
>     "outDir": "dist",
>     "lib": [
>       "dom",
>       "esnext",
>       "esnext.promise"
>     ],
>     "noUnusedLocals": true,
>     "noUnusedParameters": true,
>     "noImplicitReturns": true,
>     "noFallthroughCasesInSwitch": true,
>     "forceConsistentCasingInFileNames": true,
>     "inlineSourceMap": true,
>   }
> }
> ```
> 
> Any feedback would be much appreciated.

same with me. typescript 3.9.7

**Comment by sandersn** [comment]
@mt-yannick @alongite what VM are you using? I looked at https://caniuse.com/#search=Promise.any and it's only on recent versions of Chrome and Firefox as far as I can tell. It's not in Node 14 yet.

**Comment by mt-yannick** [comment]
@sandersn Oh! Thank you for clarifying!! So if I understand correctly, this PR implements the types for `Promise.any`, and specifying `"esnext.promise"` in `compilerOptions.libs` will make the compiler aware that `Promise.any` exists, but we are still responsible for assuring that an implementation exists!

**Comment by sandersn** [comment]
@mt-yannick yes, that's exactly right. Stuff in `"esnext"` is going to be very new, so you may need to provide your own polyfill.

**Pull request merged**